module/abci.go

```diff
@@ -117,27 +128,57 @@ func (am AppModule) handleBeforeJailedValidators(ctx context.Context) error {
 			return err
 		}
 
-		if err := sk.DeleteValidatorByPowerIndex(ctx, val); err != nil {
+		consBz, err := val.GetConsAddr()
+		if err != nil {
 			return err
 		}
 
-		// TODO: If this is used here, it persist ABCI Updates. When removes, it looks like the validator gets slashed every block in x/staking? (when we do the hack and force set jailed = false)
-		// if err := sk.DeleteLastValidatorPower(ctx, valAddr); err != nil {
-		// 	return err
-		// }
-
-		// !IMPORTANT HACK: Set validator from jailed to not jailed to see what happens
-		// Okay so this like kind of worked for a split second
-		// Issue: the validator keeps trying to be converted to a jailed validator every single block when x/staking is calling it
-		val.Jailed = false
-		if err := sk.SetValidator(ctx, val); err != nil {
+		si, err := am.keeper.GetSlashingKeeper().GetValidatorSigningInfo(ctx, consBz)
+		if err != nil {
 			return err
 		}
 
-		// remove it from persisting
-		if err := am.keeper.BeforeJailedValidators.Remove(ctx, valOperAddr); err != nil {
-			return err
+		if height == curHeight {
+			// if si.JailedUntil.After(bt) is false, then we remove from the keeper store (false positive for the val modification wrt jailing from staking hook)
+			if !si.JailedUntil.After(bt) {
+				logger.Error("EndBlocker BeforeJailedValidators validator was not jailed, removing from cache",
+					"height", height, "blockHeight", curHeight, "si.JailedUntil", si.JailedUntil, "block_time", bt, "operator", valOperAddr,
+				)
+				if err := am.keeper.CheckForJailedValidators.Remove(ctx, valOperAddr); err != nil {
+					return err
+				}
+
+				continue
+			}
+
+			logger.Error("handleBeforeJailedValidators deleting jailed validator", "height", height, "blockHeight", curHeight)
+
+			if err := sk.DeleteValidatorByPowerIndex(ctx, val); err != nil {
+				return err
+			}
+			if err := sk.DeleteLastValidatorPower(ctx, valAddr); err != nil {
+				return err
+			}
```

Comment:
Since we delete, after the validator unjails does the SDK grab the tokens, convert to BFT power & set? or does it try and use last power index?

Need to test